src/main/java/au/org/ala/names/util/DwcaWriter.java

```diff
@@ -0,0 +1,413 @@
+        addDefaultValue(coreRowType, term, defaultValue);
+    }
+
+    /**
+     * Add a default value to a term of the provided rowType.
+     *
+     * @param rowType
+     * @param term
+     * @param defaultValue
+     */
+    public void addDefaultValue(Term rowType, Term term, String defaultValue){
+
+        if(!defaultValues.containsKey(rowType)){
+            defaultValues.put(rowType, new HashMap<Term, String>());
+        }
+        Map<Term,String> currentDefaultValues= defaultValues.get(rowType);
+        if(currentDefaultValues.containsKey(term)){
+            throw new IllegalStateException("The default value of term "+ term + " is already defined");
+        }
+        currentDefaultValues.put(term, defaultValue);
+    }
+
+    /**
+     * @return new map of all current data file names by their rowTypes.
+     */
+    public Map<Term, String> getDataFiles() {
+        return Maps.newHashMap(dataFileNames);
+    }
+
+    protected void addExtensionRecord(Term rowType, Map<Term, String> row, boolean detatched) throws IOException {
+        // make sure we know the extension rowtype
+        if (!terms.containsKey(rowType)) {
+            addRowType(rowType, detatched);
+        }
+
+        // make sure we know all terms
+        List<Term> knownTerms = terms.get(rowType);
+        final boolean isFirst = knownTerms.isEmpty();
+        for (Term term : row.keySet()) {
+            if (!knownTerms.contains(term)) {
+                if (useHeaders && !isFirst){
+                    throw new IllegalStateException("You cannot add new terms after the first row when headers are enabled");
+                }
+                knownTerms.add(term);
+            }
+        }
+
+        // write extension record
+        writeRow(row, rowType, detatched);
+    }
+
+    /**
+     * Add an extension record associated with the current core record.
+     *
+     * @param rowType
+     * @param row
+     * @throws IOException
+     */
+    public void addExtensionRecord(Term rowType, Map<Term, String> row) throws IOException {
+        this.addExtensionRecord(rowType, row, false);
+    }
+
+
+    /**
+     * Add an extension record not associated with a core record.
+     *
+     * @param rowType
+     * @param row
+     * @throws IOException
+     */
+    public void addDetatchedRecord(Term rowType, Map<Term, String> row) throws IOException {
+        this.addExtensionRecord(rowType, row, true);
+    }
+
+
+    public void setEml(Dataset eml) {
+        this.eml = eml;
+    }
+
+    /**
+     * Writes meta.xml and eml.xml to the archive and closes tab writers.
+     *
+     * @deprecated Use {@link #close()} instead. This method will be removed in version 1.12.
+     */
+    @Deprecated
+    public void finalize() throws IOException {
+        close();
+    }
+
+    /**
+     * Writes meta.xml and eml.xml to the archive and closes tab writers.
+     */
+    public void close() throws IOException {
+        addEml();
```

Comment:
Encapsulate these calls in try-finally to ensure that the cleanup completes as far as possible.